app/rdf/query-cube-preview.ts

```diff
@@ -0,0 +1,278 @@
+    const scaleType = getScaleType(qScaleType?.o);
+    const qDataType = qsDim.find((q) => q.p.equals(ns.sh.datatype));
+    const qUnit = qsDim.find((q) => q.p.equals(ns.qudt.unit));
+    const qUnitLabel = qs.find(
+      (q) => q.s.equals(qUnit?.o) && q.p.equals(ns.schema.name)
+    );
+    const qDataKind = qsDim.find((q) => q.p.equals(ns.cube("meta/dataKind")));
+    const qDataKindType = qs.find(
+      (q) => q.s.equals(qDataKind?.o) && q.p.equals(ns.rdf.type)
+    );
+    const qTimeUnitType = qs.find(
+      (q) => q.s.equals(qDataKind?.o) && q.p.equals(ns.time.unitType)
+    );
+    const qIsCurrency = qs.find(
+      (q) => q.s.equals(qUnit?.o) && q.p.equals(ns.qudt.CurrencyUnit)
+    );
+    const qCurrencyExponent = qs.find(
+      (q) => q.s.equals(qUnit?.o) && q.p.equals(ns.qudt.currencyExponent)
+    );
+    const isKeyDimension = qsType.some((q) => q.o.equals(ns.cube.KeyDimension));
+    const isMeasureDimension = qsType.some((q) =>
+      q.o.equals(ns.cube.MeasureDimension)
+    );
+
+    const baseComponent: BaseComponent = {
+      cubeIri: iri,
+      iri: dimIri,
+      label: qLabel?.o.value ?? "",
+      description: qDesc?.o.value,
+      scaleType,
+      unit: qUnitLabel?.o.value,
+      order: parseNumericalTerm(qOrder?.o),
+      isNumerical: false,
+      isKeyDimension,
+      values: [],
+    };
+
+    if (isMeasureDimension) {
+      const isDecimal = qDataType?.o.equals(ns.xsd.decimal) ?? false;
+      const result: Measure = {
+        ...baseComponent,
+        __typename: resolveMeasureType(scaleType),
+        isCurrency: qIsCurrency ? true : false,
+        isDecimal,
+        currencyExponent: parseNumericalTerm(qCurrencyExponent?.o),
+        resolution: parseResolution(qDataType?.o),
+        isNumerical: getIsNumerical(qDataType?.o),
+      };
+
+      measures.push(result);
+    } else {
+      const dimensionType = resolveDimensionType(
+        getDataKind(qDataKindType?.o),
+        scaleType,
+        []
+      );
+      const baseDimension: BaseDimension = baseComponent;
+
+      switch (dimensionType) {
+        case "TemporalDimension": {
+          const timeUnit = getTimeUnit(qTimeUnitType?.o);
+          const timeFormat = getTimeFormat(qDataType?.o);
+
+          if (!timeFormat || !timeUnit) {
+            throw new Error(
+              `Temporal dimension ${dimIri} has no timeFormat or timeUnit!`
+            );
+          }
+
+          const dimension: TemporalDimension = {
+            ...baseDimension,
+            __typename: dimensionType,
+            timeFormat,
+            timeUnit,
+          };
+          dimensions.push(dimension);
+          break;
+        }
+        default: {
+          const dimension: Exclude<Dimension, TemporalDimension> = {
+            ...baseDimension,
+            __typename: dimensionType,
+          };
+          dimensions.push(dimension);
+        }
+      }
+    }
+  });
+
+  const qsObs = qs.filter(({ p }) => p.equals(ns.cube.observation));
```

Comment:
observations can also be filtered by `o.equals(ns.cube.Observation)` - which will relax the SPARQL query.
- you don't need to reconstruct the `?observationSet cube:observation ?observation` 
- thus can avoid to query `?observationSet` by stepping over ist using `cube:observationSet/cube:observation`
- which avoids duplicates and lowers the post query filtering

Reply:
Thanks for pointing this our @Rdataflow, I think we still need to have `uniqBy` to get unique observations, as they would be duplicated by the number of dimensions they have, but we can drop `?observationSet cube:observation ?observation` 👍

Reply:
@bprusinowski in order to get unique results the SPARQL query needs to be reworked first... 😄